File: kora-app-annotation-processor/src/test/java/ru/tinkoff/kora/kora/app/annotation/processor/ComponentTest.java
package ru.tinkoff.kora.kora.app.annotation.processor;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class ComponentTest extends AbstractKoraAppTest {
    @Test
    public void testComponentAnnotatedClass() {
        var draw = compile("""
            @KoraApp
            public interface ExampleApplication {
                @Root
                default Object root(TestClass object) { return java.util.Objects.requireNonNull(object); }
            }
            """, """
            @Component
            public class TestClass {
            }
            """);
        assertThat(draw.getNodes()).hasSize(2);
        draw.init().block();
    }

    @Test
    public void testComponentAnnotatedAbstractClass() {
        var draw = compile("""
            @KoraApp
            public interface ExampleApplication {
                @Root
                default Object root(TestClass object, TestInterface testInterface) { return java.util.Objects.requireNonNull(object); }
            }
            """, """
            @Component
            public abstract class TestClass {
            }
            """, """
            @Component
            public interface TestInterface {
            }
            """, """
            @Component
            public class TestClass1 extends TestClass {
            }
            """, """
            @Component
            public class TestClass2 implements TestInterface {
            }
            """);
        assertThat(draw.getNodes()).hasSize(3);
        draw.init().block();
    }
}

File: kora-app-symbol-processor/src/test/kotlin/ru/tinkoff/kora/kora/app/ksp/ComponentTest.kt
package ru.tinkoff.kora.kora.app.ksp

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test

class ComponentTest : AbstractKoraAppProcessorTest() {
    @Test
    fun testComponent() {
        val draw = compile(
            """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun test(testClass: TestClass) = ""
            }
            """.trimIndent(),
            """
            @Component
            open class TestClass()
            """.trimIndent()
        )
        Assertions.assertThat(draw.nodes).hasSize(2)
        draw.init().block()
    }

    @Test
    fun testAbstractComponent() {
        val draw = compile(
            """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun test(testClass: TestClass, testInterface: TestInterface) = ""
            }
            """.trimIndent(),
            """
            @Component
            abstract class TestClass()
            """.trimIndent(),
            """
            @Component
            interface TestInterface {}
            """.trimIndent(),
            """
            @Component
            class TestClass1() : TestClass()
            """.trimIndent(),
            """
            @Component
            class TestClass2() : TestInterface
            """.trimIndent()
        )
        Assertions.assertThat(draw.nodes).hasSize(3)
        draw.init().block()
    }
}
